delocate/libsana.py

```diff
@@ -48,54 +252,131 @@ def tree_libs(start_path, filt_func=None):
     * http://matthew-brett.github.io/pydagogue/mac_runtime_link.html
+
+    .. deprecated:: 0.9
+        This function does not support `@loader_path` and only returns the
+        direct dependencies of the libraries in `start_path`.
     """
-    lib_dict = {}
-    env_var_paths = get_environment_variable_paths()
+    warnings.warn(
+        "tree_libs doesn't support @loader_path and has been deprecated.",
+        DeprecationWarning,
+        stacklevel=2,
+    )
+    lib_dict = {}  # type: Dict[Text, Dict[Text, Text]]
     for dirpath, dirnames, basenames in os.walk(start_path):
         for base in basenames:
-            depending_libpath = realpath(pjoin(dirpath, base))
-            if filt_func is not None and not filt_func(depending_libpath):
+            depending_path = realpath(pjoin(dirpath, base))
+            if filt_func and not filt_func(depending_path):
+                logger.debug("Ignoring dependencies of: %s", depending_path)
                 continue
-            rpaths = get_rpaths(depending_libpath)
-            search_paths = rpaths + env_var_paths
-            for install_name in get_install_names(depending_libpath):
-                # If the library starts with '@rpath' we'll try and resolve it
-                # We'll do nothing to other '@'-paths
-                # Otherwise we'll search for the library using env variables
-                if install_name.startswith('@rpath'):
-                    lib_path = resolve_rpath(install_name, search_paths)
-                elif install_name.startswith('@'):
-                    lib_path = install_name
-                else:
-                    lib_path = search_environment_for_lib(install_name)
-                if lib_path in lib_dict:
-                    lib_dict[lib_path][depending_libpath] = install_name
-                else:
-                    lib_dict[lib_path] = {depending_libpath: install_name}
+            for dependency_path, install_name in get_dependencies(
+                depending_path
+            ):
+                if dependency_path is None:
+                    # Mimic deprecated behavior.
+                    # A lib_dict with unresolved paths is unsuitable for
+                    # delocating, this is a missing dependency.
+                    dependency_path = realpath(install_name)
+                if install_name.startswith("@loader_path/"):
+                    # Support for `@loader_path` would break existing callers.
+                    logger.debug(
+                        "Excluding %s because it has '@loader_path'.",
+                        install_name
+                    )
+                    continue
+                lib_dict.setdefault(dependency_path, {})
+                lib_dict[dependency_path][depending_path] = install_name
     return lib_dict
 
 
+def resolve_dynamic_paths(lib_path, rpaths, loader_path, executable_path=None):
+    # type: (Text, Iterable[Text], Optional[Text], Optional[Text]) -> Text
```

Comment:
`loader_path` no longer optional?

Reply:
It's important that `loader_path` has a correct value.  Making it optional causes issues that would be difficult to debug.  I made it optional before to be compatible with the previous function, but now that it's its own function it will no longer be optional.

An optional `loader_path` or `executable_path` would cause paths in a libraries rpaths to become invalid.  This causes issues with the order of the search paths.  I've been trying to avoid silently skipping search paths.

Reply:
Sorry - I wasn't clear - was only trying to point out that the type checking didn't match any more - totally agree about making it not-optional.